packages/esbuild-plugin-pnp/sources/index.ts

```diff
@@ -5,6 +5,41 @@ import * as fs
 const matchAll = /()/;
 const defaultExtensions = [`.tsx`, `.ts`, `.jsx`, `.mjs`, `.cjs`, `.js`, `.css`, `.json`];
 
+type External = string | {prefix: string, suffix: string};
+
+// Reference: https://github.com/evanw/esbuild/blob/537195ae84bee1510fac14235906d588084c39cd/pkg/api/api_impl.go#L366-L388
+function parseExternals(externals: Array<string>): Array<External> {
+  return externals.map(external => {
+    // ESBuild's validation pass runs before this function is called so there's no need to assert that there's a single wildcard
+    const wildcardIdx = external.indexOf(`*`);
+    if (wildcardIdx !== -1)
+      return {prefix: external.slice(0, wildcardIdx), suffix: external.slice(wildcardIdx + 1)};
+
+    return external;
+  });
+}
+
+function isExternal(path: string, externals: Array<External>): boolean {
+  for (const external of externals) {
+    if (typeof external === `object`) {
+      // Reference: https://github.com/evanw/esbuild/blob/537195ae84bee1510fac14235906d588084c39cd/internal/resolver/resolver.go#L372-L381
+      if (
+        path.length >= external.prefix.length + external.suffix.length
+        && path.startsWith(external.prefix)
+        && path.endsWith(external.suffix)
+      ) {
+        return true;
+      }
+    } else {
+      if (path === external) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
```

Comment:
Seems like this kind of thing should be utilities provided by the core rather than left to each plugin to reimplement (cc @evanw in case you're open to feedback like this)